File: cli/assets/templates/zig/src/wasm4.zig
//
// WASM-4: https://wasm4.org/docs

const std = @import("std");

fn assert_equal(a: anytype, b: anytype, comptime error_msg: []const u8) void {
    if (a != b) @compileError(std.fmt.comptimePrint("{} != {}  {s}", .{a, b, error_msg}));
}

comptime {
    const builtin = @import("builtin");
    const native_endian = builtin.target.cpu.arch.endian();
    assert_equal(native_endian, .Little, "Bit flags need little endian");
}

// ┌───────────────────────────────────────────────────────────────────────────┐
// │                                                                           │
// │ Platform Constants                                                        │
// │                                                                           │
// └───────────────────────────────────────────────────────────────────────────┘

pub const SCREEN_SIZE: u32 = 160;

// ┌───────────────────────────────────────────────────────────────────────────┐
// │                                                                           │
// │ Memory Addresses                                                          │
// │                                                                           │
// └───────────────────────────────────────────────────────────────────────────┘

pub const Memory = packed struct {
    _padding: [4]u8,
    palette: [4]Color,
    colors: DrawColors,
    gamepads: [4]GamePad,
    mouse: Mouse,
    system: SystemFlags,
    _reserved: [128]u8,
    framebuffer: [6400]u8,
    userdata: [58976]u8,
};

comptime {
    assert_equal(@bitSizeOf(Memory), 64 * 1024 * 8, "Memory layout wrong");
}

pub const m = @intToPtr(*allowzero Memory, 0);

pub const Color = packed struct {
    blue  : u8,
    green : u8,
    red   : u8,
    _reserved: u8 = 0,
};

pub const DrawColors = packed struct {
    _0: ColorIndex,
    _1: ColorIndex,
    _2: ColorIndex,
    _3: ColorIndex,
};

pub const ColorIndex = enum(u4) {
    transparent = 0,
    /// Use palette[0]
    p0   = 1,
    /// Use palette[1]
    p1   = 2,
    /// Use palette[2]
    p2   = 3,
    /// Use palette[3]
    p3   = 4,
};

pub const GamePad = packed struct {
    x: bool,
    z: bool,
    _reserved: u2,
    left: bool,
    right: bool,
    up: bool,
    down: bool,
};

pub const Mouse = packed struct {
    x: i16,
    y: i16,
    /// primary button
    b0: bool,
    /// secondary button
    b1: bool,
    /// third button
    b2: bool,
    _reserved: u5,
};

pub const SystemFlags = packed struct {
    preserve_framebuffer: bool,
    hide_gamepad_overlay: bool,
    _reserved: u6,
};

// ┌───────────────────────────────────────────────────────────────────────────┐
// │                                                                           │
// │ Raw Functions                                                             │
// │                                                                           │
// └───────────────────────────────────────────────────────────────────────────┘

const raw_api = struct {
    extern fn blit(sprite: [*]const u8, x: u32, y: u32, width: u32, height: u32, flags: u32) void;
    extern fn blitSub(sprite: [*]const u8, x: u32, y: u32, width: u32, height: u32, src_x: u32, src_y: u32, stride: u32, flags: u32) void;
    extern fn tone(frequency: u32, duration: u32, volume: u32, flags: u32) void;
};

// ┌───────────────────────────────────────────────────────────────────────────┐
// │                                                                           │
// │ Drawing Functions                                                         │
// │                                                                           │
// └───────────────────────────────────────────────────────────────────────────┘

pub const BlitFlags = packed struct {
    /// one or two bits per pixel
    two_bits:  bool = false,
    flip_x: bool = false,
    flip_y: bool = false,
    rotate: bool = false,
    _reserved: u28 = 0,
};

/// Copies pixels to the framebuffer.
pub fn blit(sprite: []const u8, x: u32, y: u32, width: u32, height: u32, flags: BlitFlags) void {
    
    raw_api.blit(sprite.ptr, x, y, width, height, @bitCast(u32, flags));
}

/// Copies a subregion within a larger sprite atlas to the framebuffer.
/// srcX: Source X position of the sprite region.
/// srcY: Source Y position of the sprite region.
/// stride: Total width of the overall sprite atlas. This is typically larger than width.
pub fn blitSub(sprite: []const u8, x: u32, y: u32, width: u32, height: u32, src_x: u32, src_y: u32, stride: u32, flags: BlitFlags) void {
    raw_api.blitSub(sprite.ptr, x, y, width, height, src_x, src_y, stride, @bitCast(u32, flags));
}

/// Draws a line between two points.
pub extern fn line(x1: u32, y1: u32, x2: u32, y2: u32) void;

/// Draws an oval (or circle).
pub extern fn oval(x: u32, y: u32, width: u32, height: u32) void;

/// Draws a rectangle.
pub extern fn rect(x: u32, y: u32, width: u32, height: u32) void;

/// Draws text using the built-in system font.
pub fn text(str: []const u8, x: u32, y: u32) void {
    textUtf8(str.ptr, str.len, x, y);
}
extern fn textUtf8(strPtr: [*]const u8, strLen: usize, x: u32, y: u32) void;

/// Draws a vertical line
pub extern fn vline(x: u32, y: u32, len: u32) void;

/// Draws a horizontal line
pub extern fn hline(x: u32, y: u32, len: u32) void;

// ┌───────────────────────────────────────────────────────────────────────────┐
// │                                                                           │
// │ Sound Functions                                                           │
// │                                                                           │
// └───────────────────────────────────────────────────────────────────────────┘

/// Plays a sound tone.
/// frequency: Wave frequency in hertz.
/// duration: Duration of the tone in frames (1/60th of a second), up to 255 frames.
/// volume: Volume of the sustain and attack durations, between 0 and 100.
pub fn tone(frequency: Tone.Frequency, duration: Tone.Duration, volume: Tone.Volume, flags: Tone.Flags) void {
    std.debug.assert(volume.is_valid());
    tone(@bitCast(u32, frequency), @bitCast(u32, duration), @bitCast(u32, volume), @bitCast(u32, flags));
}

pub const Tone = struct {
    /// Wave frequency in hertz.
    pub const Frequency = packed struct {
        start: u16,
        end: u16 = 0,
    };
    
    /// Duration of ADSR of note (unit: frames)
    ///
    ///          ^
    ///         / \ decay
    /// attack /   \ 
    ///       /     \--------
    ///      /       sustain \ release
    ///     /                 \
    pub const Duration = packed struct {
        sustain : u8 = 0,
        release : u8 = 0,
        decay   : u8 = 0,
        attack  : u8 = 0,
    };
    
    /// Volume of note (0 to 100)
    ///
    ///          ^  <-- attack volume
    ///         / \
    ///        /   \ 
    ///       /     \--------  <-- sustain valume
    ///      /               \
    ///     /                 \
    pub const Volume = packed struct {
        sustain : u8 = 100,
        attack  : u8 = 100,

        pub fn is_valid(volume: @This()) bool {
            return (0 <= volume.sustain) and (volume.sustain <= 100)
               and (0 <= volume.attack ) and (volume.attack  <= 100);
        }
    };

    pub const Flags = packed struct {
        channel: Channel,
        pulse_duty: DutyCycle,
    };

    pub const Channel = enum(u2) {
        pulse0 = 0,
        pulse1 = 1,
        triangle = 2,
        noise = 3,
    };

    pub const DutyCycle = enum(u2) {
        @"1/8" = 0,
        @"1/4" = 1,
        @"1/2" = 2,
        @"3/4" = 3, 
    };
};

pub const TONE_MODE1: u32 = 0;
pub const TONE_MODE2: u32 = 4;
pub const TONE_MODE3: u32 = 8;
pub const TONE_MODE4: u32 = 12;
pub const TONE_PAN_LEFT: u32 = 16;
pub const TONE_PAN_RIGHT: u32 = 32;

// ┌───────────────────────────────────────────────────────────────────────────┐
// │                                                                           │
// │ Storage Functions                                                         │
// │                                                                           │
// └───────────────────────────────────────────────────────────────────────────┘

/// Reads up to `size` bytes from persistent storage into the pointer `dest`.
pub extern fn diskr(dest: [*]u8, size: u32) u32;

/// Writes up to `size` bytes from the pointer `src` into persistent storage.
pub extern fn diskw(src: [*]const u8, size: u32) u32;

// ┌───────────────────────────────────────────────────────────────────────────┐
// │                                                                           │
// │ Other Functions                                                           │
// │                                                                           │
// └───────────────────────────────────────────────────────────────────────────┘

/// Prints a message to the debug console.
pub fn trace(x: []const u8) void {
    traceUtf8(x.ptr, x.len);
}
extern fn traceUtf8(strPtr: [*]const u8, strLen: usize) void;

File: cli/assets/templates/nim/src/cart/wasm4.nim
const
  SCREEN_SIZE* = 160

const
  PALETTE* = (cast[ptr array[4, uint32]](0x04))
  DRAW_COLORS* = (cast[ptr uint16](0x14))
  GAMEPAD1* = (cast[ptr uint8](0x16))
  GAMEPAD2* = (cast[ptr uint8](0x17))
  GAMEPAD3* = (cast[ptr uint8](0x18))
  GAMEPAD4* = (cast[ptr uint8](0x19))
  MOUSE_X* = (cast[ptr int16](0x1a))
  MOUSE_Y* = (cast[ptr int16](0x1c))
  MOUSE_BUTTONS* = (cast[ptr uint8](0x1e))
  SYSTEM_FLAGS* = (cast[ptr uint8](0x1f))
  NETPLAY* = (cast[ptr uint8](0x20))
  FRAMEBUFFER* = (cast[ptr array[6400, uint8]](0xa0))

const
  BUTTON_1* = 1
  BUTTON_2* = 2
  BUTTON_ENTER* = 4
  BUTTON_UNUSED* = 8
  BUTTON_LEFT* = 16
  BUTTON_RIGHT* = 32
  BUTTON_UP* = 64
  BUTTON_DOWN* = 128

const
  MOUSE_LEFT* = 0
  MOUSE_RIGHT* = 1
  MOUSE_MIDDLE* = 2

const
  BLIT_2BPP* = 1
  BLIT_1BPP* = 0
  BLIT_FLIP_X* = 2
  BLIT_FLIP_Y* = 4
  BLIT_ROTATE* = 8

const
  TONE_PULSE1* = 0
  TONE_PULSE2* = 1
  TONE_TRIANGLE* = 2
  TONE_NOISE* = 3
  TONE_MODE1* = 0
  TONE_MODE2* = 4
  TONE_MODE3* = 8
  TONE_MODE4* = 12
  TONE_PAN_LEFT* = 16
  TONE_PAN_RIGHT* = 32
  TONE_NOTE_MODE* = 64

{.push importc, codegenDecl: "__attribute__((import_name(\"$2\"))) $1 $2$3".}
proc blit*(data: ptr uint8; x: int32; y: int32; width: uint32; height: uint32;
          flags: uint32)
## Copies pixels to the framebuffer.

proc blitSub*(data: ptr uint8; x: int32; y: int32; width: uint32;
             height: uint32; srcX: uint32; srcY: uint32; stride: uint32;
             flags: uint32)
## Copies a subregion within a larger sprite atlas to the framebuffer.


proc line*(x1: int32; y1: int32; x2: int32; y2: int32)
## Draws a line between two points.

proc hline*(x: int32; y: int32; len: uint32)
## Draws a horizontal line.

proc vline*(x: int32; y: int32; len: uint32)
## Draws a vertical line.

proc oval*(x: int32; y: int32; width: uint32; height: uint32)
## Draws an oval (or circle).

proc rect*(x: int32; y: int32; width: uint32; height: uint32)
## Draws a rectangle.

proc text*(text: cstring; x: int32; y: int32)
## Draws text using the built-in system font.

proc tone*(frequency: uint32; duration: uint32; volume: uint32; flags: uint32)
## Plays a sound tone.

proc diskr*(dest: pointer; size: uint32): uint32


proc diskw*(src: pointer; size: uint32): uint32
## Writes up to `size` bytes from the pointer `src` into persistent storage.


proc trace*(str: cstring)
## Prints a message to the debug console.
proc tracef*(str: cstring) {.varargs.}
## Prints a message to the debug console.

{.pop.}

import std/macros

macro exportWasm*(def: untyped): untyped = 
  result = def
  result[^3] = nnkPragma.newTree( 
    ident("exportc"),
    nnkExprColonExpr.newTree(
      ident("codegenDecl"),
      newStrLitNode("__attribute__((export_name(\"$2\"))) $1 $2$3")
    )
  )
